Cannot build libgcrypt on osx 10.12 (probably others) due to -O2 passed in via libtool

## Issue description
Cannot build libgcrypt (or libgpg-error not quite sure which is to blame) on darwin due to optimization level -O2 on latest staging. This probably exists on other systems as well.


### Steps to reproduce
Update to latest staging, try to nix-env -p libgcrypt. On 10.12 I get the following:
```
`echo /nix/store/s9p6zi3cm0f7wvwijgwp37kb0ra5a2br-bash-4.4-p12/bin/bash ../libtool  --tag=CC   --mode=compile clang -DHAVE_CONFIG_H -I. -I..  -I../src -I../src  -I/nix/store/sfa30kvjfw5crpk2z9d715xvwc93v63n-libgpg-error-1.27-dev/include -g -O2 -Wall -c ./rndjent.c | sed -e 's/-O\([1-9s][1-9s]*\)/-O0/' -e 's/-Ofast/-O0/g' `
libtool: compile:  clang -DHAVE_CONFIG_H -I. -I.. -I../src -I../src -I/nix/store/sfa30kvjfw5crpk2z9d715xvwc93v63n-libgpg-error-1.27-dev/include -g -O0 -Wall -c ./rndjent.c  -fno-common -DPIC -o .libs/rndjent.o
/nix/store/s9p6zi3cm0f7wvwijgwp37kb0ra5a2br-bash-4.4-p12/bin/bash ../libtool  --tag=CC   --mode=compile clang -DHAVE_CONFIG_H -I. -I..  -I../gl -I../gl  -I/nix/store/sfa30kvjfw5crpk2z9d715xvwc93v63n-libgpg-error-1.27-dev/include -g -O2 -Wall -Wcast-align -Wshadow -Wstrict-prototypes -Wpointer-arith -Wno-pointer-sign -c -o der-encoder.lo der-encoder.c
In file included from ./rndjent.c:86:
./jitterentropy-base.c:53:13: warning: unknown pragma ignored [-Wunknown-pragmas]
#pragma GCC optimize ("O0")
            ^
./jitterentropy-base.c:60:4: error: "The CPU Jitter random number generator must not be compiled with optimizations. See documentation. Use the compiler switch -O0 for compiling jitterentropy-base.c."
  #error "The CPU Jitter random number generator must not be compiled with optimizations. See documentation. Use the compiler switch -O0 for compiling jitterentropy-base.c."
   ^
./jitterentropy-base.c:119:18: warning: incompatible pointer types passing 'uint64_t *' (aka 'unsigned long long *') to parameter of type 'u64 *' (aka 'unsigned long *') [-Wincompatible-pointer-types]
        jent_get_nstime(&time);
                        ^~~~~
```

I tried setting CFLAGS = "-O0" in the definition but I can't seem to get over what appears to be hardening flags setting -O2. I'm not sure how this could ever build in those circumstances.

Curious how I can override CFLAGS for this package to work around this?

## Technical details

* System: (NixOS: `nixos-version`, Ubuntu/Fedora: `lsb_release -a`, ...) $ sw_vers                                                                                                               
ProductName:	Mac OS X
ProductVersion:	10.12.6
BuildVersion:	16G29
* Nix version: (run `nix-env --version`) nix-env (Nix) 1.11.9
* Nixpkgs version: (run `nix-instantiate --eval '<nixpkgs>' -A lib.nixpkgsVersion`) "17.09pre-git"
* Sandboxing enabled: (run `grep build-use-sandbox /etc/nix/nix.conf`) n/a

--- a/pkgs/development/libraries/libgcrypt/default.nix
+++ b/pkgs/development/libraries/libgcrypt/default.nix
@@ -1,4 +1,4 @@
-{ lib, stdenv, fetchurl, libgpgerror, enableCapabilities ? false, libcap }:
+{ stdenv, fetchurl, libgpgerror, enableCapabilities ? false, libcap }:
 
 assert enableCapabilities -> stdenv.isLinux;
 
@@ -14,9 +14,13 @@ stdenv.mkDerivation rec {
   outputs = [ "out" "dev" "info" ];
   outputBin = "dev";
 
-  buildInputs =
-    [ libgpgerror ]
-    ++ lib.optional enableCapabilities libcap;
+  # The CPU Jitter random number generator must not be compiled with
+  # optimizations and the optimize -O0 pragma only works for gcc.
+  # The build enables -O2 by default for everything else.
+  hardeningDisable = stdenv.lib.optional stdenv.cc.isClang "fortify";
+
+  buildInputs = [ libgpgerror ]
+    ++ stdenv.lib.optional enableCapabilities libcap;
 
   # Make sure libraries are correct for .pc and .la files
   # Also make sure includes are fixed for callers who don't use libgpgcrypt-config

